clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -1482,6 +1477,11 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
   } else
     llvm_unreachable("no definition for emitted function");
 
+  // This is checked after emitting the function body so we know if there
+  // are any permitted infinite loops.
```

Comment:
Is this comment actually accurate?  I think the relevant code changed in 6c312954

Reply:
Comment was added in https://github.com/llvm/llvm-project/commit/ac73b73c16526c9e51943759ea6cab285a57e33f. Perhaps something along these lines could work better?
```cpp
// This is checked after emitting the function body, so as to ensure
// that the function adheres to the forward progress guarantee,
// which is required by certain optimizations.
```

Reply:
Fundamentally, I'm not sure there's any reason we need to do this after emitting the function body.  When the comment was written, it was referring to the FnIsMustProgress bit in CodeGenFunction, but that was removed.

Reply:
Seems like we fail to with some optimizations (e.g., https://github.com/llvm/llvm-project/issues/69833) due to a few non-mustprogress function, when they should likely not be.

Reply:
No, I mean, there isn't a reason to specifically do it *after* emitting the function body, as opposed to before we emit it (when we compute most other attributes).

Reply:
Not sure about this, I assumed that this was the only way to know, as per comment says, if there are any permitted infinite loops.

Reply:
If you look at the implementation of checkIfFunctionMustProgress(), it's pretty easy to verify that it doesn't actually look at the body of the function

Reply:
Anticipated before function body generation, thanks.

Reply:
Confirm this closes https://github.com/llvm/llvm-project/issues/69833.